File: src/extras-kio/libkfilefilter.cpp
#include <KFileFilter>
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <kfilefilter.h>
#include "libkfilefilter.h"
#include "libkfilefilter.hxx"

KFileFilter* KFileFilter_new() {
    return new KFileFilter();
}

KFileFilter* KFileFilter_new2(const libqt_string label, const libqt_list /* of libqt_string */ filePatterns, const libqt_list /* of libqt_string */ mimePatterns) {
    QString label_QString = QString::fromUtf8(label.data, label.len);
    QList<QString> filePatterns_QList;
    filePatterns_QList.reserve(filePatterns.len);
    libqt_string* filePatterns_arr = static_cast<libqt_string*>(filePatterns.data);
    for (size_t i = 0; i < filePatterns.len; ++i) {
        QString filePatterns_arr_i_QString = QString::fromUtf8(filePatterns_arr[i].data, filePatterns_arr[i].len);
        filePatterns_QList.push_back(filePatterns_arr_i_QString);
    }
    QList<QString> mimePatterns_QList;
    mimePatterns_QList.reserve(mimePatterns.len);
    libqt_string* mimePatterns_arr = static_cast<libqt_string*>(mimePatterns.data);
    for (size_t i = 0; i < mimePatterns.len; ++i) {
        QString mimePatterns_arr_i_QString = QString::fromUtf8(mimePatterns_arr[i].data, mimePatterns_arr[i].len);
        mimePatterns_QList.push_back(mimePatterns_arr_i_QString);
    }
    return new KFileFilter(label_QString, filePatterns_QList, mimePatterns_QList);
}

KFileFilter* KFileFilter_new3(const KFileFilter* other) {
    return new KFileFilter(*other);
}

void KFileFilter_OperatorAssign(KFileFilter* self, const KFileFilter* other) {
    self->operator=(*other);
}

bool KFileFilter_OperatorEqual(const KFileFilter* self, const KFileFilter* other) {
    return (*self == *other);
}

libqt_string KFileFilter_Label(const KFileFilter* self) {
    QString _ret = self->label();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_list /* of libqt_string */ KFileFilter_FilePatterns(const KFileFilter* self) {
    QList<QString> _ret = self->filePatterns();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of libqt_string */ KFileFilter_MimePatterns(const KFileFilter* self) {
    QList<QString> _ret = self->mimePatterns();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_string KFileFilter_ToFilterString(const KFileFilter* self) {
    QString _ret = self->toFilterString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

bool KFileFilter_IsEmpty(const KFileFilter* self) {
    return self->isEmpty();
}

bool KFileFilter_IsValid(const KFileFilter* self) {
    return self->isValid();
}

KFileFilter* KFileFilter_FromMimeType(const libqt_string mimeType) {
    QString mimeType_QString = QString::fromUtf8(mimeType.data, mimeType.len);
    return new KFileFilter(KFileFilter::fromMimeType(mimeType_QString));
}

libqt_list /* of KFileFilter* */ KFileFilter_FromMimeTypes(const libqt_list /* of libqt_string */ mimeTypes) {
    QList<QString> mimeTypes_QList;
    mimeTypes_QList.reserve(mimeTypes.len);
    libqt_string* mimeTypes_arr = static_cast<libqt_string*>(mimeTypes.data);
    for (size_t i = 0; i < mimeTypes.len; ++i) {
        QString mimeTypes_arr_i_QString = QString::fromUtf8(mimeTypes_arr[i].data, mimeTypes_arr[i].len);
        mimeTypes_QList.push_back(mimeTypes_arr_i_QString);
    }
    QList<KFileFilter> _ret = KFileFilter::fromMimeTypes(mimeTypes_QList);
    // Convert QList<> from C++ memory to manually-managed C memory
    KFileFilter** _arr = static_cast<KFileFilter**>(malloc(sizeof(KFileFilter*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new KFileFilter(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void KFileFilter_Delete(KFileFilter* self) {
    delete self;
}

File: src_original/extras-kio/libkfilefilter.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/kfilefilter.html
pub const kfilefilter = struct {
    /// New constructs a new KFileFilter object.
    ///
    ///
    pub fn New() QtC.KFileFilter {
        return qtc.KFileFilter_new();
    }

    /// New2 constructs a new KFileFilter object.
    ///
    /// ``` label: []const u8, filePatterns: [][]const u8, mimePatterns: [][]const u8, allocator: std.mem.Allocator ```
    pub fn New2(label: []const u8, filePatterns: [][]const u8, mimePatterns: [][]const u8, allocator: std.mem.Allocator) QtC.KFileFilter {
        const label_str = qtc.libqt_string{
            .len = label.len,
            .data = label.ptr,
        };
        var filePatterns_arr = allocator.alloc(qtc.libqt_string, filePatterns.len) catch @panic("kfilefilter.: Memory allocation failed");
        defer allocator.free(filePatterns_arr);
        for (filePatterns, 0..filePatterns.len) |item, i| {
            filePatterns_arr[i] = .{
                .len = item.len,
                .data = item.ptr,
            };
        }
        const filePatterns_list = qtc.libqt_list{
            .len = filePatterns.len,
            .data = filePatterns_arr.ptr,
        };
        var mimePatterns_arr = allocator.alloc(qtc.libqt_string, mimePatterns.len) catch @panic("kfilefilter.: Memory allocation failed");
        defer allocator.free(mimePatterns_arr);
        for (mimePatterns, 0..mimePatterns.len) |item, i| {
            mimePatterns_arr[i] = .{
                .len = item.len,
                .data = item.ptr,
            };
        }
        const mimePatterns_list = qtc.libqt_list{
            .len = mimePatterns.len,
            .data = mimePatterns_arr.ptr,
        };

        return qtc.KFileFilter_new2(label_str, filePatterns_list, mimePatterns_list);
    }

    /// New3 constructs a new KFileFilter object.
    ///
    /// ``` other: QtC.KFileFilter ```
    pub fn New3(other: ?*anyopaque) QtC.KFileFilter {
        return qtc.KFileFilter_new3(@ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/kfilefilter.html#operator-eq)
    ///
    /// ``` self: QtC.KFileFilter, other: QtC.KFileFilter ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.KFileFilter_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/kfilefilter.html#operator-eq-eq)
    ///
    /// ``` self: QtC.KFileFilter, other: QtC.KFileFilter ```
    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.KFileFilter_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/kfilefilter.html#label)
    ///
    /// ``` self: QtC.KFileFilter, allocator: std.mem.Allocator ```
    pub fn Label(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KFileFilter_Label(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kfilefilter.Label: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kfilefilter.html#filePatterns)
    ///
    /// ``` self: QtC.KFileFilter, allocator: std.mem.Allocator ```
    pub fn FilePatterns(self: ?*anyopaque, allocator: std.mem.Allocator) [][]const u8 {
        const _arr: qtc.libqt_list = qtc.KFileFilter_FilePatterns(@ptrCast(self));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("kfilefilter.FilePatterns: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("kfilefilter.FilePatterns: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kfilefilter.html#mimePatterns)
    ///
    /// ``` self: QtC.KFileFilter, allocator: std.mem.Allocator ```
    pub fn MimePatterns(self: ?*anyopaque, allocator: std.mem.Allocator) [][]const u8 {
        const _arr: qtc.libqt_list = qtc.KFileFilter_MimePatterns(@ptrCast(self));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("kfilefilter.MimePatterns: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("kfilefilter.MimePatterns: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kfilefilter.html#toFilterString)
    ///
    /// ``` self: QtC.KFileFilter, allocator: std.mem.Allocator ```
    pub fn ToFilterString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KFileFilter_ToFilterString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kfilefilter.ToFilterString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kfilefilter.html#isEmpty)
    ///
    /// ``` self: QtC.KFileFilter ```
    pub fn IsEmpty(self: ?*anyopaque) bool {
        return qtc.KFileFilter_IsEmpty(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kfilefilter.html#isValid)
    ///
    /// ``` self: QtC.KFileFilter ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.KFileFilter_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kfilefilter.html#fromMimeType)
    ///
    /// ``` mimeType: []const u8 ```
    pub fn FromMimeType(mimeType: []const u8) QtC.KFileFilter {
        const mimeType_str = qtc.libqt_string{
            .len = mimeType.len,
            .data = mimeType.ptr,
        };
        return qtc.KFileFilter_FromMimeType(mimeType_str);
    }

    /// [Qt documentation](https://api.kde.org/kfilefilter.html#fromMimeTypes)
    ///
    /// ``` mimeTypes: [][]const u8, allocator: std.mem.Allocator ```
    pub fn FromMimeTypes(mimeTypes: [][]const u8, allocator: std.mem.Allocator) []QtC.KFileFilter {
        var mimeTypes_arr = allocator.alloc(qtc.libqt_string, mimeTypes.len) catch @panic("kfilefilter.FromMimeTypes: Memory allocation failed");
        defer allocator.free(mimeTypes_arr);
        for (mimeTypes, 0..mimeTypes.len) |item, i| {
            mimeTypes_arr[i] = .{
                .len = item.len,
                .data = item.ptr,
            };
        }
        const mimeTypes_list = qtc.libqt_list{
            .len = mimeTypes.len,
            .data = mimeTypes_arr.ptr,
        };
        const _arr: qtc.libqt_list = qtc.KFileFilter_FromMimeTypes(mimeTypes_list);
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.KFileFilter, _arr.len) catch @panic("kfilefilter.FromMimeTypes: Memory allocation failed");
        const _data: [*]QtC.KFileFilter = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kfilefilter.html#dtor.KFileFilter)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.KFileFilter ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KFileFilter_Delete(@ptrCast(self));
    }
};
